relay/circuit-v2.md

```diff
@@ -0,0 +1,398 @@
+However, the protocol reserves the slot to support the functionality for the rare cases where it is actually desirable to use active relay functionality in a controlled environment.
+
+If the relay has a reservation (and thus an active connection) from the peer, then it opens the second hop of the connection using the `stop` protocol; the details are not relevant for the `hop` protocol and the only thing that matters is whether it succeeds in opening the relay connection or not.
+If the relayed connection is successfully established, then the relay responds with `HopMessage` with `type = STATUS` and `status = OK`:
+```
+HopMessage {
+  type = STATUS
+  sgtatus = OK
+  limit = Limit {...}
+}
+```
+at this point the original `hop` stream becomes the relayed connection.
+The `limit` field, if present, communicates to the initiator the
+limits applied to the relayed connection with the semantics described
+[above](#reservation).
+
+If the relayed connection cannot be established, then the relay responds with a `HopMessage` of `type = STATUS` and the `status` field having a value other than `OK`.
+Common failure status codes are:
+- `PERMISSION_DENIED` if the connection is rejected because of peer filtering using ACLs.
+- `NO_RESERVATION` if there is no active reservation for the target peer
+- `RESOURCE_LIMIT_EXCEEDED` if there are two many relayed connections from the initiator or to the target peer.
+- `CONNECTION_FAILED` if the relay failed to terminate the connection to the target peer.
+
+
+### Stop Protocol
+
+The Stop protocol governs connection termination between the relay and the target peer;
+it uses the protocol ID `/libp2p/circuit/relay/0.2.0/stop`.
+
+In order to terminate a relayed connection, the relay opens a stream
+using an existing connection to the target peer. If there is no
+existing connection, an active relay may attempt to open one using the
+initiator supplied address, but as discussed in the previous section
+this functionality is generally deprecated.
+
+The relay sends a `StopMessage` with `type = CONNECT` and the following form:
+```
+StopMessage {
+  type = CONNECT
+  peer = Peer { ID = ...}
+  limit = Limit { ...}
+}
+```
+- the `peer` field contains a `Peer` struct with the peer `ID` of the connection initiator.
+- the `limit` field, if present, conveys the limits applied to the relayed connection with the semantics described [above](#reservation).
+
+If the target peer accepts the connection it responds to the relay with a `StopMessage` of `type = STATUS` and `status = OK`:
+```
+StopMessage {
+  type = STATUS
+  status = OK
+}
+```
+
+If the target fails to terminate the connection for some reason, then it responds to the relay with a `StopMessage` of `type = STATUS` and the `status` code set to something other than `OK`.
+Common failure status codes are:
+- `CONNECTION_FAILED` if the target internally failed to create the relayed connection for some reason.
+
```

Comment:
Should we also note that the A <=> B connection takes over the stream at this point? Also noting the point above (full connection establishment flow runs).

Reply:
it's implied, but it doesn't hurt to explicitly note it.